include/seastar/coroutine/generator.hh

```diff
@@ -457,4 +449,320 @@ public:
+    }
+
+    yield_awaiter final_suspend() noexcept {
+        _slice = nullptr;
+        return yield_awaiter{this, this->_consumer};
+    }
+
+    void unhandled_exception() noexcept {
+        _exception = std::current_exception();
+    }
+
+    yield_awaiter yield_value(Yielded slice) noexcept {
+        // an empty slice is forbidden, otherwise the increment_awaiter would resume
+        // from suspension, but end up finding nothing to return to the consumer.
+        assert(!std::ranges::empty(slice));
+        this->_slice = std::addressof(slice);
+        return yield_awaiter{this, this->_consumer};
+    }
+
+    void return_void() noexcept {}
+
+    // @return if the generator has reached the end of the sequence
+    bool finished() const noexcept {
+        return _slice == nullptr;
+    }
+
+    // @return the current slice produced by the producer
+    Yielded& slice() noexcept {
+        assert(_slice);
+        return *_slice;
+    }
+
+
+    void rethrow_if_unhandled_exception() {
+        if (_exception) {
+            std::rethrow_exception(std::move(_exception));
+        }
+    }
+
+    void run_and_dispose() noexcept final {
+        using handle_type = std::coroutine_handle<generator_promise_base>;
+        handle_type::from_promise(*this).resume();
+    }
+
+    seastar::task* waiting_task() noexcept final {
+        return _waiting_task;
+    }
+
+private:
+    friend class next_awaiter<Yielded>;
+};
+
+template <typename Yielded>
+class next_awaiter {
+protected:
+    generator_promise_base<Yielded>* _promise = nullptr;
+    std::coroutine_handle<> _producer = nullptr;
+
+public:
+    explicit next_awaiter(std::nullptr_t) noexcept {}
+    next_awaiter(generator_promise_base<Yielded>& promise,
+                 std::coroutine_handle<> producer) noexcept
+        : _promise{std::addressof(promise)}
+        , _producer{producer} {}
+
+    template <typename Promise>
+    std::coroutine_handle<> await_suspend(std::coroutine_handle<Promise> consumer) noexcept {
+        _promise->_consumer = consumer;
+        return _producer;
+    }
+};
+
+template<typename Ref, typename Value, typename Yielded>
+class [[nodiscard]] generator {
+    using value_type = std::conditional_t<std::is_void_v<Value>,
+                                          std::remove_cvref_t<Ref>,
+                                          Value>;
+    using reference_type = std::conditional_t<std::is_void_v<Value>,
+                                              Ref&&,
+                                              Ref>;
+    using yielded_type = Yielded;
+
+public:
+    class promise_type;
+
+private:
+    using handle_type = std::coroutine_handle<promise_type>;
+    handle_type _coro = {};
+
+public:
+    class iterator;
+
+    generator() noexcept = default;
+    explicit generator(promise_type& promise) noexcept
+        : _coro(std::coroutine_handle<promise_type>::from_promise(promise))
+    {}
+
+    generator(const generator&) = delete;
```

Comment:
What about the `generator(generator&&)` ctor?
Shouldn't it be defined? (`= default`?)

Reply:
indeed! no wonder i could not reuse the `verify_fib_drained()` for testing the buffered generator. thanks for pointing this out!